tempesta_fw/http.c

```diff
@@ -223,23 +223,85 @@ tfw_http_prep_304(TfwHttpMsg *resp, TfwHttpReq *req, void *msg_it,
 	if (tfw_http_msg_setup(resp, it, data_len))
 		return TFW_BLOCK;
 
-	tfw_http_msg_write(it, resp, &rh);
+	ret = tfw_http_msg_write(it, resp, &rh);
 	if (end)
-		tfw_http_msg_write(it, resp, end);
+		ret |= tfw_http_msg_write(it, resp, end);
 
 	TFW_DBG("Send HTTP 304 response\n");
 
-	return TFW_PASS;
+	return ret ? TFW_BLOCK : TFW_PASS;
+}
+
+/*
+ * Free an HTTP message.
+ * Also, free the connection instance if there's no more references.
+ *
+ * This function should be used anytime when there's a chance that
+ * a connection instance may belong to multiple messages, which is
+ * almost always. If a connection is suddenly closed then it still
+ * can be safely dereferenced and used in the code.
+ * In rare cases we're sure that a connection instance in a message
+ * doesn't have multiple users. For example, when an error response
+ * is prepared and sent by Tempesta, that HTTP message does not need
+ * a connection instance. The message is then immediately destroyed,
+ * and a simpler tfw_http_msg_free() can be used for that.
+ *
+ * NOTE: @hm->conn may be NULL if @hm is the response that was served
+ * from cache.
+ */
+static void
+tfw_http_conn_msg_free(TfwHttpMsg *hm)
+{
+	if (unlikely(!hm))
+		return;
+
+	if (hm->conn) {
+		/*
+		 * Unlink the connection while there is at least one
+		 * reference. Use atomic exchange to avoid races with
+		 * new messages arriving on the connection.
+		 */
+		__cmpxchg((unsigned long *)&hm->conn->msg, (unsigned long)hm,
+			  0UL, sizeof(long));
+		tfw_connection_put(hm->conn);
+	}
+
+	tfw_http_msg_free(hm);
+}
+
+/*
+ * Close the client connection and free unpaired request. This function
+ * is needed for cases when we cannot prepare response for this request.
+ * As soon as request is not linked with any response, sending to that
+ * client stops starting with that request, because that creates
+ * a "hole" in the chain of requests -- a request without a response.
+ * Subsequent responses cannot be sent to the client until that
+ * hole is closed, which at this point will never happen. To solve
+ * this situation there is no choice but to close client connection.
+ */
+void
+tfw_http_resp_build_error(TfwHttpReq *req)
+{
+	ss_close_sync(req->conn->sk, true);
+	spin_lock(&((TfwCliConn *)req->conn)->seq_qlock);
+	if (likely(!list_empty(&req->msg.seq_list)))
+		list_del_init(&req->msg.seq_list);
+	spin_unlock(&((TfwCliConn *)req->conn)->seq_qlock);
+	tfw_http_conn_msg_free((TfwHttpMsg *)req);
+	TFW_INC_STAT_BH(clnt.msgs_otherr);
```

Comment:
I am not convinced that removal of a request from `seq_list`, and then freeing it, is necessary here. This will work just fine, yes, but it looks like unnecessary complication (and extra unnecessary code).

My understanding is this. When a error response cannot be sent out due to an error while preparing it, a "hole" in the `seq_list`  will cause a stop in sending responses to that client. No subsequent responses will be sent to it, starting with the response to this request `req`. Then the erroneous client connection is closed. That causes the connection drop and the invocation of `tfw_http_conn_cli_drop()`, which takes care of all requests that are still in the queue, including this one.

In other words, I believe, that one the first statement in this function is required to handle the situation properly. I am not sure of the last statement, as from that point the client connection should be considered closed.

Reply:
My suspicions about request leak in this case were caused by comments and code from  `tfw_http_conn_cli_drop()`:
```
/*                                                                                                                                                                 
 * Drop client connection's resources.                                                                                                                             
 *                                                                                                                                                                 
 * Desintegrate the client connection's @seq_list. Requests that have                                                                                              
 * a paired response can be freed. Move those to @zap_queue for doing                                                                                              
 * that without the lock. Requests without a paired response have not                                                                                              
 * been answered yet. They are held in the lists of server connections                                                                                             
 * until responses come. Don't free those requests.
```
. . .
. . .

```
       spin_lock(&cli_conn->seq_qlock);
        list_for_each_entry_safe(req, tmp, seq_queue, msg.seq_list) {
                if (req->resp)
                        list_move_tail(&req->msg.seq_list, &zap_queue);
                else
                        list_del_init(&req->msg.seq_list);
        }
        spin_unlock(&cli_conn->seq_qlock);
```
As it seems - unpaired request will be removed from `seq_queue` but will not be freed - to wait the paired response. But in our case response has already been arrived - so the request will be lost.
And as we have to explicitly free request - we also have to remove it from `seq_queue` before that, otherwise - we will catch `BUG_ON` in
```
tfw_http_req_destruct(void *msg)
{
        TfwHttpReq *req = msg;

        BUG_ON(!list_empty(&req->msg.seq_list));
. . .
```

Reply:
Yes, you're absolutely correct, I missed that. Never underestimate the value of a good comment!

I have another question though. Did you consider just passing a NULL `resp` to `tfw_http_resp_fwd()`, and then handling this special case by closing the client connection? In my view, that may provide a cleaner path for handling this rare error. It might be easier to follow and understand, because it follows the same flow of execution as in the normal case. Locks are natural there (already exist), and all there's to do there is to handle a NULL `resp` in an appropriate place.

It appears that the whole needed flow is already there. For instance, you could do this:
```C
BUG_ON(list_empty(&req->msg.seq_list));
if (!resp) {
        list_del_init(&req->msg.seq_list);
        spin_unlock(&cli_conn->seq_qlock);
        ss_close_sync(cli_conn->sk, true);
        tfw_http_conn_msg_free((TfwHttpMsg *)req);
        TFW_INC_STAT_BH(serv.msgs_otherr);
        return;
}
req->resp = (TfwHttpMsg *)resp;
```

Note that the case where `seq_list` is empty is handled before this piece of code in `tfw_http_resp_fwd()`, and it will correctly handle a NULL `resp`.

There's one caveat here though. The same caveat exists with your original code that handles this error outside of `tfw_http_resp_fwd()`.

Let's say there are 10 CPUs (from 1 to 10), and there are 10 requests in `seq_list`.  Let's say this kind of error occurs on CPU №10. Now, let's say the order in which other 9 `seq_list` entries are filled with a `resp` is: 9, 8, 7, 6, 5, 4, 3, 2, 1. Until the first entry is filled with a `resp`, all of these responses are not sent out to the client.

If CPU 10 comes last after all of these, then all 9 responses are sent out to the client. If CPU 10 comes first, then none of the 9 already received and prepared responses are sent out. You get the picture.

So the question is, can we handle this special case in `tfw_http_resp_fwd()` so that we surely send all preceding (ready) responses to the client before we close the connection? Perhaps, with a special flag? I believe that for proper efficiency, this is a valid case to think about...

Reply:
As for me, passing `NULL` response to `tfw_http_resp_fwd()` as a part of error processing doesn't seem to be a clear and nice. But it's a matter of taste mostly.

Currently we have traditional approach with separate normal and error processing:
```C
/* Some actions to prepare response (resp) */
if (error_on_prepare_resp)
    goto error;

tfw_http_resp_fwd(req, resp);
return;

error:
kfree(resp);
tfw_http_resp_build_error(req);
```
But if `tfw_http_resp_build_error()` is moved to `tfw_http_resp_fwd()`, error processing will be implemented in other way:
```C
/* Some actions to prepare response (resp) */
if (error_on_prepare_resp) {
    kfree(resp);
    resp = NULL;
}

tfw_http_resp_fwd(req, resp);
return;
```
In the second case error processing is implicitly made in `tfw_http_resp_fwd(req, resp)` call and that makes the call to be mandatory. That's why i don't really like the approach. But as i said, it's matter of taste.

Concerning  second problem you have mentioned (drops of received but not forwarded responses). It's a hard problem actually. There is only one reason why Tempesta fails to prepare response on it's own: lack of memory. If single memory error happen, it's very likely that more are coming. It seems reasonable to drop one clients to save memory for others. Tempesta's behaviour is not tested thoroughly on random memory allocation errors which can potentially cause very different errors.

Reply:
I tend to agree with @ikoveshnikov that handling errors at the place they arise is more traditional, explicit, and better understood.

Per request from @ikoveshnikov, here's a sample comment that describes the case where errors are handled _outside_ of `tfw_http_resp_fwd()`. A comment like this would help to describe the intricacies of the case and help to remember the choice that has been made.

*_As a consequence of closing a client connection on error of preparing a response, it's possible that some already prepared responses will not be sent to the client. That depends on the order in which CPUs close the connection and call tfw_http_resp_fwd(). This is the intended behaviour. The goal is to free some memory at the cost of dropping a few clients, so that Tempesta can continue working._*

Reply:
@keshonok, this is an interesting case, very strange however. Basically, we can simplify it to a race between preparing response to 10th request and 1st one - all other responses are just blocked by unfinished 1st request. The problem happens only on absence of memory, i.e. response for the 1st request must be received and processed just before the problem happens and next the CPU must stuck a little bit giving 1th CPU to face memory starvation. The case should be rare and insignificant in case of race between only 2 messages. The case with all 10 messages is even less probable. Next, we typically need some memory for network transmission (keep in mind that skb clones etc), so its likely that all the 9 messages still won't be delivered. Last, all the messages are synchronized via socket work queue to send data over the socket and close it, so if we're lucky we can deliver 9 responses before closing the socket due to an error.

Memory starvation is very serious thing introducing OOM killing of various processes and failures on most of allocation calls. It's better to choose some unlucky connections/clients and completely free the resources to get enough memory instead of trying to do anything under heavy memory pressure, which leads to long scannings for free memory pages etc.

Reply:
Added specified comment for 'tfw_http_resp_build_error' function.